File: 백준/Gold/2470. 두 용액/두 용액.py
import sys
input = sys.stdin.readline
n = int(input())
l = list(map(int,input().split()))
l.sort()
s = 0
e = len(l) - 1
r = 10**10

while s < e:
    x = l[s] + l[e]
    if abs(x) < r:
        r = abs(x)
        sa,aa = l[s], l[e]
    if x < 0:
        s += 1
    else:
        e -= 1
print(sa,aa)

File: 백준/Gold/2470. 두 용액/두 용액.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br;
    static StringTokenizer st;
    static int n,res = Integer.MAX_VALUE, se,ee;
    static int[] arr;

    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        arr = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(arr);
        int start = 0;
        int end = n-1;

        while(start<end){
            int num = Math.abs(arr[start] + arr[end]);
            if(num < res){
                res = num;
                se = arr[start];
                ee = arr[end];
            }
            if(arr[start] + arr[end] < 0){
                start++;
            }else{
                end--;
            }
        }
//        print();
        System.out.println(se + " " + ee);
    }

    private static void print() {
        System.out.println(Arrays.toString(arr));
    }
}